zipkin-autoconfigure/storage-cassandra3/src/main/java/zipkin/autoconfigure/storage/cassandra3/brave/TracedSession.java

```diff
@@ -11,204 +11,182 @@
-      brave.clientSpanThreadBinder().setCurrentSpan(null);
-      return new BraveResultSetFuture(target.executeAsync(statement), brave);
+    private static final Executor EXECUTOR = MoreExecutors.directExecutor();
+    private final Session session;
+    private final Brave brave;
+    private final SpanCollector collector;
+    private final ClientTracer clientTracer;
+    private final long useKeyspaceTimeout;
+    private volatile boolean keyspaceSet = false;
+    private final ProtocolVersion version;
+
+    static TracedSession create(final Session session, Brave brave, SpanCollector collector) {
+        return new TracedSession(session, brave, collector);
     }
-    try {
-      return method.invoke(target, args);
-    } catch (InvocationTargetException e) {
-      if (e.getCause() instanceof RuntimeException) throw e.getCause();
-      throw e;
+
+    private TracedSession(final Session session, Brave brave, SpanCollector collector) {
+        this.session = session;
+        this.brave = brave;
+        this.collector = collector;
+        this.clientTracer = brave.clientTracer();
+        this.useKeyspaceTimeout = session.getCluster().getConfiguration().getSocketOptions().getConnectTimeoutMillis();
+        this.version = session.getCluster().getConfiguration().getProtocolOptions().getProtocolVersion();
     }
-  }
-
-  @Override public void update(Host host, Statement statement, Exception e, long nanos) {
-    if (!(statement instanceof BoundStatement)) return;
-    Span span = cache.remove(statement);
-    if (span == null) {
-      if (statement.isTracing()) {
-        LOG.warn("{} not in the cache eventhough tracing is on", statement);
-      }
-      return;
+
+    @Override
+    public String getLoggedKeyspace() {
+        return session.getLoggedKeyspace();
     }
-    span.setDuration(nanos / 1000); // TODO: allow client tracer to end with duration
-    Endpoint local = span.getAnnotations().get(0).host; // TODO: expose in brave
-    long endTs = span.getTimestamp() + span.getDuration();
-    span.addToAnnotations(Annotation.create(endTs, "cr", local));
-    if (e != null) {
-      span.addToBinary_annotations(BinaryAnnotation.create(Constants.ERROR, e.getMessage(), local));
+
+    @Override
+    public Session init() {
+        return session.init();
     }
-    int ipv4 = ByteBuffer.wrap(host.getAddress().getAddress()).getInt();
-    Endpoint endpoint = Endpoint.create("cassandra3", ipv4, host.getSocketAddress().getPort());
-    span.addToBinary_annotations(BinaryAnnotation.address("sa", endpoint));
-    collector.collect(span);
-  }
-
-  @Override
-  public boolean equals(Object obj) {
-    if (obj instanceof TracedSession) {
-      TracedSession other = (TracedSession) obj;
-      return target.equals(other.target);
+
+    @Override
+    public ResultSetFuture executeAsync(final Statement statement) {
+        // don't trace any already self-traced spans
```

Comment:
before, I added this for a couple reasons.. one is to address perhaps your concern about the server amplifying itself, and another is to special-case it so that we don't have as much of a test burden. What's worrying me about trying to generify this is that it is getting past the point where we should kite without tests..

Ex the below was a heavily guarded thing that ensured we don't have to worry about things like local collectors amplifying, nor any other traces we don't expect to make. A generic cassandra tracer would have to be more... generic, but then it would have a higher burden of testing etc. If that's the goal, I might copy/paste the work and then make a cassandra client tracer out of it
```
-    // Only join traces, don't start them. This prevents LocalCollector's thread from amplifying.
 -    if (brave.serverSpanThreadBinder().getCurrentServerSpan() != null &&
 -        brave.serverSpanThreadBinder().getCurrentServerSpan().getSpan() != null
 -        && method.getName().equals("executeAsync") && args[0] instanceof BoundStatement) {
```